### PR TITLE
Read API_IMPORT_PATH from api/go.mod in `make grpc`

### PR DESCRIPTION
**Pull request by codingllama**
API\_IMPORT\_PATH is consistently being resolved as an empty string, breaking proto generation.

Since the path is fixed, it seems simpler to read api/go.mod and do away with the Go program.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -841,9 +843,6 @@ buildbox-grpc:
 		lib/multiplexer/test/ping.proto \
 		lib/web/envelope.proto
 
-# we eval within the make target to avoid invoking `go run` with every other call to the makefile
-	$(eval API_IMPORT_PATH := $(shell go run build.assets/gomod/print-import-path/main.go ./api))
```
Comment (Joerger) [inline Makefile:845]: Can we instead change this to `$(eval API_IMPORT_PATH := $(shell head -1 api/go.mod | awk '{print $$2}'))` to catch when the api path is changed between major versions.

Comment (codingllama) [reply Makefile:845]: Done, also pushed variables to target scope.

**Comment by codingllama** [comment]
Example invocation:

```
$ make grpc
(...)
make: Entering directory '/go/src/github.com/gravitational/teleport'
PROTO_INCLUDE = /usr/local/include:/go/src:/go/src/github.com/gogo/protobuf/protobuf:/go/src/github.com/gogo/protobuf:/go/src/github.com/gogo/protobuf/protobuf
/usr/bin/clang-format-10 -i -style='{ColumnLimit: 100, IndentWidth: 4, Language: Proto}' \
	api/client/proto/authservice.proto \
	api/client/proto/joinservice.proto \
	api/types/events/events.proto \
	api/types/types.proto \
	api/types/webauthn/webauthn.proto \
	api/types/wrappers/wrappers.proto \
	lib/datalog/types.proto \
	lib/events/slice.proto \
	lib/multiplexer/test/ping.proto \
	lib/web/envelope.proto
cd api/client/proto && protoc -I=.:$PROTO_INCLUDE \
	--gogofast_out=plugins=grpc,Mgithub.com/gravitational/teleport/api/types/events/events.proto=github.com/gravitational/teleport/api/types/events,Mgithub.com/gravitational/teleport/api/types/types.proto=github.com/gravitational/teleport/api/types,Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=github.com/gravitational/teleport/api/types/webauthn,Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=github.com/gravitational/teleport/api/types/wrappers,Mignoreme=ignoreme:. \
	certs.proto authservice.proto joinservice.proto
cd api/types/events && protoc -I=.:$PROTO_INCLUDE \
	--gogofast_out=plugins=grpc,Mgithub.com/gravitational/teleport/api/types/events/events.proto=github.com/gravitational/teleport/api/types/events,Mgithub.com/gravitational/teleport/api/types/types.proto=github.com/gravitational/teleport/api/types,Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=github.com/gravitational/teleport/api/types/webauthn,Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=github.com/gravitational/teleport/api/types/wrappers,Mignoreme=ignoreme:. \
	events.proto
cd api/types && protoc -I=.:$PROTO_INCLUDE \
	--gogofast_out=plugins=grpc,Mgithub.com/gravitational/teleport/api/types/events/events.proto=github.com/gravitational/teleport/api/types/events,Mgithub.com/gravitational/teleport/api/types/types.proto=github.com/gravitational/teleport/api/types,Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=github.com/gravitational/teleport/api/types/webauthn,Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=github.com/gravitational/teleport/api/types/wrappers,Mignoreme=ignoreme:. \
	types.proto
cd api/types/webauthn && protoc -I=.:$PROTO_INCLUDE \
	--gogofast_out=plugins=grpc,Mgithub.com/gravitational/teleport/api/types/events/events.proto=github.com/gravitational/teleport/api/types/events,Mgithub.com/gravitational/teleport/api/types/types.proto=github.com/gravitational/teleport/api/types,Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=github.com/gravitational/teleport/api/types/webauthn,Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=github.com/gravitational/teleport/api/types/wrappers,Mignoreme=ignoreme:. \
	webauthn.proto
cd api/types/wrappers && protoc -I=.:$PROTO_INCLUDE \
	--gogofast_out=plugins=grpc,Mgithub.com/gravitational/teleport/api/types/events/events.proto=github.com/gravitational/teleport/api/types/events,Mgithub.com/gravitational/teleport/api/types/types.proto=github.com/gravitational/teleport/api/types,Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=github.com/gravitational/teleport/api/types/webauthn,Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=github.com/gravitational/teleport/api/types/wrappers,Mignoreme=ignoreme:. \
	wrappers.proto
cd lib/datalog && protoc -I=.:$PROTO_INCLUDE \
	--gogofast_out=plugins=grpc,Mgithub.com/gravitational/teleport/api/types/events/events.proto=github.com/gravitational/teleport/api/types/events,Mgithub.com/gravitational/teleport/api/types/types.proto=github.com/gravitational/teleport/api/types,Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=github.com/gravitational/teleport/api/types/webauthn,Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=github.com/gravitational/teleport/api/types/wrappers,Mignoreme=ignoreme:. \
	types.proto
cd lib/events && protoc -I=.:$PROTO_INCLUDE \
	--gogofast_out=plugins=grpc,Mgithub.com/gravitational/teleport/api/types/events/events.proto=github.com/gravitational/teleport/api/types/events,Mgithub.com/gravitational/teleport/api/types/types.proto=github.com/gravitational/teleport/api/types,Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=github.com/gravitational/teleport/api/types/webauthn,Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=github.com/gravitational/teleport/api/types/wrappers,Mignoreme=ignoreme:. \
	slice.proto
cd lib/multiplexer/test && protoc -I=.:$PROTO_INCLUDE \
	--gogofast_out=plugins=grpc,Mgithub.com/gravitational/teleport/api/types/events/events.proto=github.com/gravitational/teleport/api/types/events,Mgithub.com/gravitational/teleport/api/types/types.proto=github.com/gravitational/teleport/api/types,Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=github.com/gravitational/teleport/api/types/webauthn,Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=github.com/gravitational/teleport/api/types/wrappers,Mignoreme=ignoreme:. \
	ping.proto
cd lib/web && protoc -I=.:$PROTO_INCLUDE \
	--gogofast_out=plugins=grpc,Mgithub.com/gravitational/teleport/api/types/events/events.proto=github.com/gravitational/teleport/api/types/events,Mgithub.com/gravitational/teleport/api/types/types.proto=github.com/gravitational/teleport/api/types,Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=github.com/gravitational/teleport/api/types/webauthn,Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=github.com/gravitational/teleport/api/types/wrappers,Mignoreme=ignoreme:. \
	envelope.proto
make: Leaving directory '/go/src/github.com/gravitational/teleport'
```

**Comment by codingllama** [comment]
PTAL?

File: Makefile [hunk 2]
```diff
@@ -820,19 +820,24 @@ enter:
...
+# flag. This way protoc generated files will use the correct api module import
+# path in the case where the import path has a version suffix, e.g.
+# "github.com/gravitational/teleport/api/v8".
+buildbox-grpc: GOGOPROTO_IMPORTMAP := $\
+	Mgithub.com/gravitational/teleport/api/types/events/events.proto=$(API_IMPORT_PATH)/types/events,$\
+	Mgithub.com/gravitational/teleport/api/types/types.proto=$(API_IMPORT_PATH)/types,$\
+	Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=$(API_IMPORT_PATH)/types/webauthn,$\
+	Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=$(API_IMPORT_PATH)/types/wrappers,$\
+	Mignoreme=ignoreme
 buildbox-grpc:
```
Comment (ibeckermayer) [inline Makefile:839]: What are the semantics/purpose of having multiple `.PHONY` targets of the same name? I haven't been able to find good information on that online.

Might it work to just make this
```Makefile
# buildbox-grpc generates GRPC stubs
.PHONY: buildbox-grpc
buildbox-grpc:
	API_IMPORT_PATH := $(shell head -1 api/go.mod | awk '{print $$2}')

# Proto file dependencies within the api module must be passed with the 'M'
# flag. This way protoc generated files will use the correct api module import
# path in the case where the import path has a version suffix, e.g.
# "github.com/gravitational/teleport/api/v8".
	GOGOPROTO_IMPORTMAP := $\
	Mgithub.com/gravitational/teleport/api/types/events/events.proto=$(API_IMPORT_PATH)/types/events,$\
	Mgithub.com/gravitational/teleport/api/types/types.proto=$(API_IMPORT_PATH)/types,$\
	Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=$(API_IMPORT_PATH)/types/webauthn,$\
	Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=$(API_IMPORT_PATH)/types/wrappers,$\
	Mignoreme=ignoreme

	@echo "PROTO_INCLUDE = $$PROTO_INCLUDE"
	$(CLANG_FORMAT) -i -style='{ColumnLimit: 100, IndentWidth: 4, Language: Proto}' \
		api/client/proto/authservice.proto \
		api/client/proto/joinservice.proto \
		api/types/events/events.proto \
		api/types/types.proto \
		api/types/webauthn/webauthn.proto \
		api/types/wrappers/wrappers.proto \
		lib/datalog/types.proto \
		lib/events/slice.proto \
		lib/multiplexer/test/ping.proto \
		lib/web/envelope.proto

	cd api/client/proto && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		certs.proto authservice.proto joinservice.proto

	cd api/types/events && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		events.proto

	cd api/types && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		types.proto

	cd api/types/webauthn && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		webauthn.proto

	cd api/types/wrappers && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		wrappers.proto

	cd lib/datalog && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		types.proto

	cd lib/events && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		slice.proto

	cd lib/multiplexer/test && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		ping.proto

	cd lib/web && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		envelope.proto

```

**Review by Joerger — APPROVED** [review]

File: Makefile (hunk 2 above)
Comment (codingllama) [reply Makefile:839]: These are just per-target variable overrides. It does what your snippet is trying to do.

I don't think your snippet works, though. We can't define make-like variable in the target's body.

**Comment by codingllama** [comment]
PTAL?

File: Makefile (hunk 2 above)
Comment (Joerger) [reply Makefile:839]: It's worth noting that target-specific variables like this one are initialized on every call to `make`, whereas setting the variable within the target like Isaiah is meaning to ( which can be done with $(eval API_IMPORT_PATH := $(shell head -1 api/go.mod | awk '{print $$2}')) ) would only be initialized in the specified target. You can see this is true if you add `shell && sleep 5 ...`. I'm not sure which approach is actually best practice for that reason, though here the performance impact is negligible.

**Review by timothyb89 — APPROVED** [review]

File: Makefile (hunk 2 above)
Comment (codingllama) [reply Makefile:839]: Yep. It should be negligible in this case, so I'd rather take the cleaner syntax to the evals (which I think are less usual).

**Review by ibeckermayer — APPROVED** [review]

**Pull request merged**